java.net.SocketException: Network is unreachable

`api "com.airbnb.android:lottie:3.0.7"`

Nexus 5X
Android 7.1.2

```
java.lang.IllegalStateException: Unable to parse composition
	at com.airbnb.lottie.LottieAnimationView$2.onResult(LottieAnimationView.java:75)
	at com.airbnb.lottie.LottieAnimationView$2.onResult(LottieAnimationView.java:73)
	at com.airbnb.lottie.LottieTask.notifyFailureListeners(LottieTask.java:158)
	at com.airbnb.lottie.LottieTask.access$200(LottieTask.java:27)
	at com.airbnb.lottie.LottieTask$1.run(LottieTask.java:133)
	at android.os.Handler.handleCallback(Handler.java:751)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:154)
	at android.app.ActivityThread.main(ActivityThread.java:6121)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
	at de.robv.android.xposed.XposedBridge.main(XposedBridge.java:107)
Caused by: java.net.SocketException: Network is unreachable
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:334)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:196)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:178)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:356)
	at java.net.Socket.connect(Socket.java:605)
	at com.android.okhttp.internal.Platform.connectSocket(Platform.java:113)
	at com.android.okhttp.Connection.connectSocket(Connection.java:196)
	at com.android.okhttp.Connection.connect(Connection.java:172)
	at com.android.okhttp.Connection.connectAndSetOwner(Connection.java:367)
	at com.android.okhttp.OkHttpClient$1.connectAndSetOwner(OkHttpClient.java:130)
	at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:329)
	at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:246)
	at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:457)
	at com.android.okhttp.internal.huc.HttpURLConnectionImpl.connect(HttpURLConnectionImpl.java:126)
	at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.connect(DelegatingHttpsURLConnection.java:89)
	at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.connect(HttpsURLConnectionImpl.java)
	at com.airbnb.lottie.network.NetworkFetcher.fetchFromNetworkInternal(NetworkFetcher.java:94)
	at com.airbnb.lottie.network.NetworkFetcher.fetchFromNetwork(NetworkFetcher.java:79)
	at com.airbnb.lottie.network.NetworkFetcher.fetchSync(NetworkFetcher.java:48)
	at com.airbnb.lottie.network.NetworkFetcher.fetchSync(NetworkFetcher.java:31)
	at com.airbnb.lottie.LottieCompositionFactory$1.call(LottieCompositionFactory.java:71)
	at com.airbnb.lottie.LottieCompositionFactory$1.call(LottieCompositionFactory.java:68)
	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
	at java.lang.Thread.run(Thread.java:761)
```

--- a/lottie/src/main/java/com/airbnb/lottie/LottieAnimationView.java
+++ b/lottie/src/main/java/com/airbnb/lottie/LottieAnimationView.java
@@ -33,13 +33,9 @@ import com.airbnb.lottie.value.SimpleLottieValueCallback;
 
 import java.io.ByteArrayInputStream;
 import java.io.InputStream;
-import java.net.SocketException;
-import java.net.UnknownHostException;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
-
-import javax.net.ssl.SSLException;
 
 import static com.airbnb.lottie.RenderMode.HARDWARE;
 
@@ -69,8 +65,9 @@ import static com.airbnb.lottie.RenderMode.HARDWARE;
   private static final LottieListener<Throwable> DEFAULT_FAILURE_LISTENER = new LottieListener<Throwable>() {
     @Override public void onResult(Throwable throwable) {
       // By default, fail silently for network errors.
-      if (throwable instanceof UnknownHostException || throwable instanceof SocketException || throwable instanceof SSLException) {
+      if (Utils.isNetworkException(throwable)) {
         Logger.warning("Unable to load composition.", throwable);
+        return;
       }
       throw new IllegalStateException("Unable to parse composition", throwable);
     }

--- a/lottie/src/main/java/com/airbnb/lottie/LottieAnimationView.java
+++ b/lottie/src/main/java/com/airbnb/lottie/LottieAnimationView.java
@@ -11,21 +11,21 @@ import android.graphics.drawable.Drawable;
 import android.os.Build;
 import android.os.Parcel;
 import android.os.Parcelable;
+import android.text.TextUtils;
+import android.util.AttributeSet;
+import android.util.Log;
+import android.view.View;
+
 import androidx.annotation.FloatRange;
 import androidx.annotation.MainThread;
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 import androidx.annotation.RawRes;
 import androidx.appcompat.widget.AppCompatImageView;
-import okio.Okio;
-
-import android.text.TextUtils;
-import android.util.AttributeSet;
-import android.util.Log;
-import android.view.View;
 
 import com.airbnb.lottie.model.KeyPath;
 import com.airbnb.lottie.parser.moshi.JsonReader;
+import com.airbnb.lottie.utils.Logger;
 import com.airbnb.lottie.utils.Utils;
 import com.airbnb.lottie.value.LottieFrameInfo;
 import com.airbnb.lottie.value.LottieValueCallback;
@@ -33,13 +33,15 @@ import com.airbnb.lottie.value.SimpleLottieValueCallback;
 
 import java.io.ByteArrayInputStream;
 import java.io.InputStream;
-import java.io.StringReader;
+import java.net.SocketException;
+import java.net.UnknownHostException;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
 
+import javax.net.ssl.SSLException;
+
 import static com.airbnb.lottie.RenderMode.HARDWARE;
-import static com.airbnb.lottie.RenderMode.SOFTWARE;
 
 /**
  * This view will load, deserialize, and display an After Effects animation exported with
@@ -64,6 +66,15 @@ import static com.airbnb.lottie.RenderMode.SOFTWARE;
 @SuppressWarnings({"unused", "WeakerAccess"}) public class LottieAnimationView extends AppCompatImageView {
 
   private static final String TAG = LottieAnimationView.class.getSimpleName();
+  private static final LottieListener<Throwable> DEFAULT_FAILURE_LISTENER = new LottieListener<Throwable>() {
+    @Override public void onResult(Throwable throwable) {
+      // By default, fail silently for network errors.
+      if (throwable instanceof UnknownHostException || throwable instanceof SocketException || throwable instanceof SSLException) {
+        Logger.warning("Unable to load composition.", throwable);
+      }
+      throw new IllegalStateException("Unable to parse composition", throwable);
+    }
+  };
 
   private final LottieListener<LottieComposition> loadedListener = new LottieListener<LottieComposition>() {
     @Override public void onResult(LottieComposition composition) {
@@ -71,11 +82,7 @@ import static com.airbnb.lottie.RenderMode.SOFTWARE;
     }
   };
 
-  private final LottieListener<Throwable> failureListener = new LottieListener<Throwable>() {
-    @Override public void onResult(Throwable throwable) {
-      throw new IllegalStateException("Unable to parse composition", throwable);
-    }
-  };
+  private LottieListener<Throwable> failureListener = DEFAULT_FAILURE_LISTENER;
 
   private final LottieDrawable lottieDrawable = new LottieDrawable();
   private boolean isInitialized;
@@ -371,6 +378,34 @@ import static com.airbnb.lottie.RenderMode.SOFTWARE;
    */
   public void setAnimationFromUrl(String url) {
     setCompositionTask(LottieCompositionFactory.fromUrl(getContext(), url));
+  }
+
+  /**
+   * Set a default failure listener that will be called if any of the setAnimation APIs fail for any reason.
+   * This can be used to replace the default behavior which is to crash.
+   *
+   * @see #resetFailureListener()
+   */
+  public void setFailureListener(LottieListener<Throwable> failureListener) {
+    if (compositionTask != null) {
+      compositionTask.removeFailureListener(this.failureListener);
+      compositionTask.addFailureListener(failureListener);
+    }
+    this.failureListener = failureListener;
+  }
+
+  /**
+   * Clears the failure listener set with {@link #setFailureListener(LottieListener)} and restores the default behavior
+   * which is to crash.
+   */
+  public void resetFailureListener() {
+    if (failureListener == DEFAULT_FAILURE_LISTENER) {
+      return;
+    }
+    if (compositionTask != null) {
+      compositionTask.removeFailureListener(failureListener);
+    }
+    this.failureListener = DEFAULT_FAILURE_LISTENER;
   }
 
   private void setCompositionTask(LottieTask<LottieComposition> compositionTask) {

--- a/lottie/src/main/java/com/airbnb/lottie/utils/Utils.java
+++ b/lottie/src/main/java/com/airbnb/lottie/utils/Utils.java
@@ -22,6 +22,19 @@ import com.airbnb.lottie.animation.content.TrimPathContent;
 import com.airbnb.lottie.animation.keyframe.FloatKeyframeAnimation;
 
 import java.io.Closeable;
+import java.io.InterruptedIOException;
+import java.net.BindException;
+import java.net.ConnectException;
+import java.net.NoRouteToHostException;
+import java.net.PortUnreachableException;
+import java.net.ProtocolException;
+import java.net.SocketException;
+import java.net.SocketTimeoutException;
+import java.net.UnknownHostException;
+import java.net.UnknownServiceException;
+import java.nio.channels.ClosedChannelException;
+
+import javax.net.ssl.SSLException;
 
 public final class Utils {
   public static final int SECOND_IN_NANOS = 1000000000;
@@ -240,6 +253,16 @@ public final class Utils {
     Bitmap resizedBitmap = Bitmap.createScaledBitmap(bitmap, width, height, true);
     bitmap.recycle();
     return resizedBitmap;
+  }
+
+  /**
+   * From http://vaibhavblogs.org/2012/12/common-java-networking-exceptions/
+   */
+  public static boolean isNetworkException(Throwable e) {
+    return e instanceof SocketException || e instanceof ClosedChannelException ||
+        e instanceof InterruptedIOException || e instanceof ProtocolException ||
+        e instanceof SSLException || e instanceof UnknownHostException ||
+        e instanceof UnknownServiceException;
   }
 
   /**

